src/hotspot/share/utilities/concurrentHashTable.inline.hpp

```diff
@@ -1298,4 +1339,75 @@ inline bool ConcurrentHashTable<CONFIG, F>::
   return true;
 }
 
+template <typename CONFIG, MEMFLAGS F>
+class ConcurrentHashTable<CONFIG, F>::BucketsClaimer {
+  // Default size of _claim_size_log2
+  static const size_t DEFAULT_CLAIM_SIZE_LOG2 = 7;
+
+  ConcurrentHashTable<CONFIG, F>* _cht;
+  // The table is split into ranges, every increment is one range.
+  volatile size_t _next_to_claim;
+  size_t _claim_size_log2; // Log number of buckets in claimed range.
+  size_t _limit;      // Limit to number of claims
+
+  // If there is a paused resize, we also need to operate on the already resized items.
+  volatile size_t _next_to_claim_new_table;
+  size_t _claim_size_log2_new_table;
+  size_t _limit_new_table;
+
+public:
+  BucketsClaimer(ConcurrentHashTable<CONFIG, F>* cht) :
+    _cht(cht),
+    _next_to_claim(0),
+    _claim_size_log2(DEFAULT_CLAIM_SIZE_LOG2),
+    _limit(0),
+    _next_to_claim_new_table(0),
+    _claim_size_log2_new_table(0),
+    _limit_new_table(0)
+  {
+    size_t size_log2 = _cht->_table->_log2_size;
+    _claim_size_log2 = MIN2(_claim_size_log2, size_log2);
+    _limit = (size_t)1 << (size_log2 - _claim_size_log2);
```

Comment:
Is it really advantageous to keep these values as log2? It seems to me that just storing the actual non-log values would actually require less work everywhere.
I am fine with restricting claim increments to powers of two (but even that seems artificial), there only seems to be no advantage to use logs anywhere (only that every use needs to do the shift to get the actual value).